backend/cmd/headlamp.go

```diff
@@ -132,6 +136,77 @@ func StartHeadlampServer(config *HeadlampConfig) {
 
 	// Configuration
 	r.HandleFunc("/config", clientConf.getConfig).Methods("GET")
+	r.HandleFunc("/authtype", getClusterAuthType).Queries("cluster", "{cluster}").Methods("GET")
+
+	state := "app"
+	var oauthConfig oauth2.Config
+	var verifier *oidc.IDTokenVerifier
+	ctx := context.Background()
+
+	r.HandleFunc("/oidc", func(w http.ResponseWriter, r *http.Request) {
+		cluster := r.URL.Query().Get("cluster")
+		authConfig, err := GetClusterAuthConfig(cluster)
+		if err != nil {
+			log.Fatal(err)
+		}
+		provider, err := oidc.NewProvider(ctx, authConfig.IdpIssuerURL)
+		if err != nil {
+			log.Fatal(err)
+		}
+		oidcConfig := &oidc.Config{
+			ClientID: authConfig.ClientID,
+		}
+		verifier = provider.Verifier(oidcConfig)
+		oauthConfig = oauth2.Config{
+			ClientID:     authConfig.ClientID,
+			ClientSecret: authConfig.ClientSecret,
+			Endpoint:     provider.Endpoint(),
+			RedirectURL:  fmt.Sprintf("http://127.0.0.1:%1s/callback", config.port),
+			Scopes:       []string{oidc.ScopeOpenID, "profile", "email", "groups"},
+		}
+		http.Redirect(w, r, oauthConfig.AuthCodeURL(state), http.StatusFound)
+	}).Queries("cluster", "{cluster}")
+
+	r.HandleFunc("/callback", func(w http.ResponseWriter, r *http.Request) {
+		if r.URL.Query().Get("state") != state {
+			http.Error(w, "state did not match", http.StatusBadRequest)
+			return
+		}
+		oauth2Token, err := oauthConfig.Exchange(ctx, r.URL.Query().Get("code"))
+		if err != nil {
+			http.Error(w, "Failed to exchange token: "+err.Error(), http.StatusInternalServerError)
+			return
+		}
+		rawIDToken, ok := oauth2Token.Extra("id_token").(string)
+		if !ok {
+			http.Error(w, "No id_token field in oauth2 token.", http.StatusInternalServerError)
+			return
+		}
+		idToken, err := verifier.Verify(ctx, rawIDToken)
+		if err != nil {
+			http.Error(w, "Failed to verify ID Token: "+err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		resp := struct {
+			OAuth2Token   *oauth2.Token
+			IDTokenClaims *json.RawMessage // ID Token payload is just JSON.
+		}{oauth2Token, new(json.RawMessage)}
+
+		if err := idToken.Claims(&resp.IDTokenClaims); err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		var redirectUrl string
+		if config.devMode {
+			redirectUrl = "http://localhost:3000/"
+		} else {
+			redirectUrl = fmt.Sprintf("http://localhost:%1s/", config.port)
+		}
+		redirectUrl += "auth?token=" + oauth2Token.AccessToken
```

Comment:
This has no information about the cluster for which we're getting the token. So if the user starts authenticating in a cluster, then switches to a different one, and in the popup window completes the authentication, I am guessing it results in the frontend setting up the token for the wrong cluster.

Reply:
hmm, we will have to store on the server info about which cluster has started the auth flow(this can be stored as a global property which will be changed by an api call) and than in the callback we can pass the current auth asking cluster in the query param